source/mir/blas/dot.d

```diff
@@ -0,0 +1,127 @@
+/**
+License: $(LINK2 http://boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Authors: Ilya Yaroshenko
+*/
+module mir.blas.dot;
+
+import mir.ndslice.slice;
+import std.traits;
+
+/++
+Params:
+	x = vector
+	y = vector
+Returns:
+	scalar `xᵀ × y`
+Note:
+	`dot` implementation is naive for now.
++/
+auto dot(V1, V2)(scope V1 x, scope V2 y)
+	if ((isDynamicArray!V1 || is(V1 : Slice!(1, T1*), T1)) &&
+		(isDynamicArray!V2 || is(V2 : Slice!(1, T2*), T2)) &&
+		is(Unqual!(ForeachType!V1) == Unqual!(ForeachType!V2)))
+in
+{
+	assert(x.length == y.length);
+}
+body
+{
+	alias T = Unqual!(ForeachType!V1);
+	static if(isDynamicArray!V1 && isDynamicArray!V2)
+	{
+		return dotImpl(x, y);
+	}
+	else
+	static if(isDynamicArray!V1 && !isDynamicArray!V2)
+	{
+		import mir.internal.utility;
+		return
+			y.stride == 1 ?
+				dotImpl(y.toDense, x) :
+				dotImpl(x, y);
+	}
+	else
+	static if(!isDynamicArray!V1 && isDynamicArray!V2)
+	{
+		import mir.internal.utility;
+		return
+			x.stride == 1 ?
+				dotImpl(x.toDense, y) :
+				dotImpl(y, x);
+	}
```

Comment:
Can't you combine those two cases with alias?

alias f1 = V1, f2 = V2;

static if(isDynamicArray!V1 && !isDynamicArray!V2)
{
   f1 = V2, f2= V1;
}
static if(!isDynamicArray!V1 && isDynamicArray!V2)
{
                import mir.internal.utility;
 		return
 			x.stride == 1 ?
 				dotImpl(x.toDense, y) :
 				dotImpl(y, x);
}